jsonschema/_validators.py

```diff
@@ -190,9 +190,14 @@ def enum(validator, enums, instance, schema):
 
 
 def ref(validator, ref, instance, schema):
-    with validator.resolver.resolving(ref) as resolved:
+    scope, resolved = validator.resolver.resolve(ref)
+    validator.resolver.push_scope(scope)
```

Comment:
I did experiment with using a class with `__enter__` and `__exit__` and it was about the same as `contextlib.contextmanager` (possibly even a little slower).  It's always possible I was doing something non-optimal.